// @irys/sdk/build/cjs/web/tokens/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRedstonePrice = void 0;
const tslib_1 = require("tslib");
const utils_1 = require("../utils");
const base64url_1 = tslib_1.__importDefault(require("base64url"));
const axios_1 = tslib_1.__importDefault(require("axios"));
const utils_2 = tslib_1.__importDefault(require("../../common/utils"));
class BaseWebToken {
    constructor(config) {
        this.minConfirm = 5;
        this.isSlow = false;
        this.needsFee = true;
        this.inheritsRPC = false;
        Object.assign(this, config);
        this.config = config;
    }
    // common methods
    get address() {
        return this._address;
    }
    ready() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            this._address = this.wallet ? this.ownerToAddress(yield this.getPublicKey()) : undefined;
        });
    }
    getId(item) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return base64url_1.default.encode(Buffer.from(yield (0, utils_1.getCryptoDriver)().hash(yield item.rawSignature())));
        });
    }
    price() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return getRedstonePrice(this.ticker);
        });
    }
}
exports.default = BaseWebToken;
function getRedstonePrice(token) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const res = yield axios_1.default.get(`https://api.redstone.finance/prices?symbol=${token}&provider=redstone&limit=1`);
        yield utils_2.default.checkAndThrow(res, "Getting price data");
        return res.data[0].value;
    });
}
exports.getRedstonePrice = getRedstonePrice;
