// @irys/sdk/build/cjs/node/flavours/arweave.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArweaveIrys = void 0;
const tslib_1 = require("tslib");
const base_1 = tslib_1.__importDefault(require("../base"));
const arweave_1 = tslib_1.__importDefault(require("../tokens/arweave"));
class ArweaveIrys extends base_1.default {
    constructor({ url, key, config }) {
        super({
            url,
            config,
            getTokenConfig: (irys) => {
                var _a;
                return new arweave_1.default({
                    irys,
                    name: "arweave",
                    ticker: "AR",
                    minConfirm: 10,
                    providerUrl: (_a = config === null || config === void 0 ? void 0 : config.providerUrl) !== null && _a !== void 0 ? _a : "https://arweave.net",
                    wallet: key,
                    isSlow: true,
                    opts: config === null || config === void 0 ? void 0 : config.tokenOpts,
                });
            },
        });
    }
}
exports.ArweaveIrys = ArweaveIrys;
