// @irys/sdk/build/cjs/node/flavours/ethereum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthereumIrys = void 0;
const tslib_1 = require("tslib");
const base_1 = tslib_1.__importDefault(require("../base"));
const ethereum_1 = tslib_1.__importDefault(require("../tokens/ethereum"));
class EthereumIrys extends base_1.default {
    constructor({ url, key, config }) {
        super({
            url,
            config,
            getTokenConfig: (irys) => {
                var _a;
                return new ethereum_1.default({
                    irys,
                    name: "ethereum",
                    ticker: "ETH",
                    providerUrl: (_a = config === null || config === void 0 ? void 0 : config.providerUrl) !== null && _a !== void 0 ? _a : "https://cloudflare-eth.com/",
                    wallet: key,
                    opts: config === null || config === void 0 ? void 0 : config.tokenOpts,
                });
            },
        });
    }
}
exports.EthereumIrys = EthereumIrys;
