// @irys/sdk/build/cjs/node/tokens/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRedstonePrice = exports.BaseNodeToken = void 0;
const tslib_1 = require("tslib");
const axios_1 = tslib_1.__importDefault(require("axios"));
const utils_1 = tslib_1.__importDefault(require("../../common/utils"));
class BaseNodeToken {
    constructor(config) {
        this.minConfirm = 5;
        this.isSlow = false;
        this.needsFee = true;
        Object.assign(this, config);
        this._address = this.wallet ? this.ownerToAddress(this.getPublicKey()) : undefined;
    }
    // common methods
    get address() {
        return this._address;
    }
    price() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return getRedstonePrice(this.ticker);
        });
    }
}
exports.BaseNodeToken = BaseNodeToken;
function getRedstonePrice(token) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const res = yield axios_1.default.get(`https://api.redstone.finance/prices?symbol=${token}&provider=redstone&limit=1`);
        yield utils_1.default.checkAndThrow(res, "Getting price data");
        return res.data[0].value;
    });
}
exports.getRedstonePrice = getRedstonePrice;
