// @irys/sdk/build/cjs/node/cli.js
#!/usr/bin/env node
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.exportForTesting = exports.program = void 0;
const tslib_1 = require("tslib");
// Note: DO NOT REMOVE/ALTER THE ABOVE LINE - it is called a 'shebang' and is vital for CLI execution.
const bignumber_js_1 = tslib_1.__importDefault(require("bignumber.js"));
const commander_1 = require("commander");
const fs_1 = require("fs");
const inquirer_1 = tslib_1.__importDefault(require("inquirer"));
const irys_1 = tslib_1.__importDefault(require("./irys"));
const upload_1 = require("./upload");
exports.program = new commander_1.Command();
let balpad, walpad; // padding state variables
// Define the CLI flags for the program
exports.program
    .option("-h, --host <string>", "Irys node hostname/URL (eg http://node1.irys.xyz)")
    .option("-n, --network <string>", "The network to use", "mainnet")
    .option("-w, --wallet <string>", "Path to keyfile or the private key itself", "default")
    .option("-t, --token <string>", "The token to use")
    .option("-c --currency <string>", "DEPRECATED: the currency to use (same as token)")
    .option("--timeout <number>", "The timeout (in ms) for API HTTP requests - increase if you get timeouts for upload")
    .option("--no-confirmation", "Disable confirmations for certain actions")
    .option("--tags [value...]", "Tags to include, format <name> <value>")
    .option("--multiplier <number>", "Adjust the multiplier used for tx rewards - the higher the faster the network will process the transaction.", "1.00")
    .option("--batch-size <number>", "Adjust the upload-dir batch size (process more items at once - uses more resources (network, memory, cpu) accordingly!)", "5")
    .option("-d, --debug", "Increases verbosity of errors and logs additional debug information. Used for troubleshooting.", false)
    .option("--index-file <string>", "Name of the file to use as an index for upload-dir manifests (relative to the path provided to upload-dir).")
    .option("--provider-url <string>", "Override the provider URL")
    .option("--contract-address <string>", "Override the contract address")
    .option("--content-type <string>", "Override the content type for *ALL* files uploaded")
    .option("--remove-deleted", "Removes previously uploaded (but now deleted) items from the manifest")
    .option("--force-chunking", "Forces usage of chunking for all files regardless of size");
// Define commands
// uses NPM view to query the package's version.
exports.program.version(irys_1.default.VERSION, "-v, --version", "Gets the current package version of the Irys client");
// Balance command - gets the provided address' balance on the specified bundler
exports.program
    .command("balance")
    .description("Gets the specified user's balance for the current Irys node")
    .argument("<address>", "address")
    .action((address) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    try {
        options.address = balpad ? address.substring(1) : address;
        const irys = yield init(options, "balance");
        const balance = yield irys.utils.getBalance(options.address);
        console.log(`Balance: ${balance} ${irys.tokenConfig.base[0]} (${irys.utils.unitConverter(balance).toFixed()} ${irys.token})`);
    }
    catch (err) {
        console.error(`Error whilst getting balance: ${options.debug ? err.stack : err.message} `);
        return;
    }
}));
// Withdraw command - sends a withdrawal request for n base units to the specified bundler for the loaded wallet
exports.program
    .command("withdraw")
    .description("Sends a fund withdrawal request")
    .argument("<amount>", "amount to withdraw in token base units")
    .action((amount) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    try {
        const irys = yield init(options, "withdraw");
        const confirmed = yield confirmation(`Confirmation: withdraw ${amount} ${irys.tokenConfig.base[0]} from ${irys.api.config.url.host} (${yield irys.utils.getBundlerAddress(irys.token)})?\n Y / N`);
        if (confirmed) {
            const res = yield irys.withdrawBalance(new bignumber_js_1.default(amount));
            console.log(`Withdrawal request for ${res === null || res === void 0 ? void 0 : res.requested} ${irys.tokenConfig.base[0]} successful\nTransaction ID: ${res === null || res === void 0 ? void 0 : res.tx_id} with network fee ${res === null || res === void 0 ? void 0 : res.fee} for a total cost of ${res === null || res === void 0 ? void 0 : res.final} `);
        }
        else {
            console.log("confirmation failed");
        }
    }
    catch (err) {
        console.error(`Error whilst sending withdrawal request: ${options.debug ? err.stack : err.message} `);
        return;
    }
}));
// Upload command - Uploads a specified file to the specified bundler using the loaded wallet.
exports.program
    .command("upload")
    .description("Uploads a specified file")
    .argument("<file>", "relative path to the file you want to upload")
    .action((file) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    try {
        const irys = yield init(options, "upload");
        const tags = parseTags(options === null || options === void 0 ? void 0 : options.tags);
        const res = yield irys.uploadFile(file, { tags: tags !== null && tags !== void 0 ? tags : [] });
        console.log(`Uploaded to https://gateway.irys.xyz/${res === null || res === void 0 ? void 0 : res.id}`);
    }
    catch (err) {
        console.error(`Error whilst uploading file: ${options.debug ? err.stack : err.message} `);
        return;
    }
}));
exports.program
    .command("upload-dir")
    .description("Uploads a folder (with a manifest)")
    .argument("<folder>", "relative path to the folder you want to upload")
    .action((folder) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    yield uploadDir(folder);
}));
// Deploy command - DEPRECATED
exports.program
    .command("deploy")
    .description("(DEPRECATED - use the functionally identical 'upload-dir' instead.) Deploys a folder (with a manifest) to the specified bundler")
    .argument("<folder>", "relative path to the folder you want to deploy")
    .action((folder) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    console.warn("WARN: Deploy is deprecated, use the functionally identical 'upload-dir' instead.");
    yield uploadDir(folder);
}));
function uploadDir(folder) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        try {
            const irys = yield init(options, "upload");
            const tags = parseTags(options === null || options === void 0 ? void 0 : options.tags);
            const res = yield irys.uploadFolder(folder, {
                indexFile: options.indexFile,
                batchSize: +options.batchSize,
                interactivePreflight: options.confirmation,
                keepDeleted: !options.removeDeleted,
                manifestTags: tags !== null && tags !== void 0 ? tags : [],
                logFunction: (log) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                    console.log(log);
                }),
                itemOptions: {
                    upload: {},
                },
            });
            if (!res)
                return console.log("Nothing to upload");
            console.log(`Uploaded to https://gateway.irys.xyz/${res.id}`);
        }
        catch (err) {
            console.error(`Error whilst uploading ${folder} - ${options.debug ? err.stack : err.message}`);
        }
    });
}
const parseTags = (arr) => {
    if (!arr)
        return;
    if (arr.length % 2 !== 0)
        throw new Error(`Tags key is missing a value!`);
    return arr.reduce((a, v, i) => {
        (i + 1) % 2 === 0 ? (a.at(-1).value = v) : a.push({ name: v, value: "" });
        return a;
    }, []);
};
exports.program
    .command("fund")
    .description("Funds your account with the specified amount of atomic units")
    .argument("<amount>", "Amount to add in atomic units")
    .action((amount) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    try {
        if (isNaN(+amount))
            throw new Error("Amount must be an integer");
        const irys = yield init(options, "fund");
        const confirmed = yield confirmation(`Confirmation: send ${amount} ${irys.tokenConfig.base[0]} (${irys.utils.unitConverter(amount).toFixed()} ${irys.token}) to ${irys.api.config.url.host} (${yield irys.utils.getBundlerAddress(irys.token)})?\n Y / N`);
        if (confirmed) {
            const tx = yield irys.fund(new bignumber_js_1.default(amount), options.multiplier);
            console.log(`Funding receipt: \nAmount: ${tx.quantity} with Fee: ${tx.reward} to ${tx.target} \nTransaction ID: ${tx.id} `);
        }
        else {
            console.log("confirmation failed");
        }
    }
    catch (err) {
        console.error(`Error whilst funding: ${options.debug ? err.stack : err.message} `);
        return;
    }
}));
exports.program
    .command("price")
    .description("Check how much of a specific token is required for an upload of <amount> bytes")
    .argument("<bytes>", "The number of bytes to get the price for")
    .action((bytes) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    try {
        if (isNaN(+bytes))
            throw new Error("Amount must be an integer");
        const irys = yield init(options, "price");
        yield irys.utils.getBundlerAddress(options.token); // will throw if the bundler doesn't support the token
        const cost = yield irys.utils.getPrice(options.token, +bytes);
        console.log(`Price for ${bytes} bytes in ${options.token} is ${cost.toFixed(0)} ${irys.tokenConfig.base[0]} (${irys.utils
            .unitConverter(cost)
            .toFixed()} ${irys.token})`);
    }
    catch (err) {
        console.error(`Error whilst getting price: ${options.debug ? err.stack : err.message} `);
        return;
    }
}));
/**
 * Interactive CLI prompt allowing a user to confirm an action
 * @param message the message specifying the action they are asked to confirm
 */
function confirmation(message) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        if (!(options === null || options === void 0 ? void 0 : options.confirmation)) {
            return true;
        }
        const answers = yield inquirer_1.default.prompt([{ type: "input", name: "confirmation", message }]);
        return answers.confirmation.toLowerCase() == "y";
    });
}
/**
 * Initialisation routine for the CLI, mainly for initialising a Irys instance
 * @param opts the parsed options from the cli
 * @returns a new Irys instance
 */
function init(opts, operation) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        let wallet;
        let irys;
        // every option needs a host/network and token so ensure they're present
        if (!(opts.host || opts.network)) {
            throw new Error("Host (-h) or network (-n) parameter is required!");
        }
        if (!opts.token) {
            throw new Error("token flag (-t, --token) is required!");
        }
        // some operations do not require a wallet
        if (!["balance", "price"].includes(operation)) {
            // require a wallet
            if (opts.wallet === "default") {
                // default to wallet.json under the right conditions
                if (opts.token === "arweave" && (yield (0, upload_1.checkPath)("./wallet.json"))) {
                    wallet = yield loadWallet("./wallet.json");
                }
                else {
                    throw new Error("Wallet (-w) required for this operation!");
                }
            }
            else {
                // remove padding if present
                wallet = yield loadWallet(walpad ? opts.wallet.substring(1) : opts.wallet);
            }
        }
        try {
            // create and ready the Irys instance
            irys = new irys_1.default({
                url: opts.host,
                network: opts.network,
                token: opts.token.toLowerCase(),
                key: wallet !== null && wallet !== void 0 ? wallet : "",
                config: {
                    providerUrl: opts.providerUrl,
                    contractAddress: opts.contractAddress,
                    timeout: opts.timeout,
                },
            });
            yield irys.ready();
        }
        catch (err) {
            throw new Error(`Error initialising Irys client - ${options.debug ? err.stack : err.message}`);
        }
        // log the loaded address
        if (wallet && irys.address) {
            console.log(`Loaded address: ${irys.address}`);
        }
        if (opts.contentType) {
            irys.uploader.contentType = opts.contentType;
        }
        if (opts.forceChunking) {
            irys.uploader.useChunking = true;
        }
        return irys;
    });
}
/**
 * Loads a wallet file from the specified path into a JWK interface
 * @param path path to the JWK file
 * @returns JWK interface
 */
function loadWallet(path) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        if (yield (0, upload_1.checkPath)(path)) {
            if (options.debug) {
                console.log("Loading wallet file");
            }
            return JSON.parse((0, fs_1.readFileSync)(path).toString());
        }
        else {
            if (options.debug) {
                console.log("Assuming raw key instead of keyfile path");
            }
            return path;
        }
    });
}
const options = exports.program.opts();
if (options.currency)
    options.token = options.currency;
const isScript = require.main === module;
if (isScript) {
    // to debug CLI: log process argv, load into var, and run in debugger.
    // console.log(JSON.stringify(process.argv));
    // process.exit(1);
    // replace this with dumped array. (make sure to append/include --no-confirmation)
    const argv = process.argv;
    // padding hack
    // this is because B64URL strings can start with a "-" which makes commander think it's a flag
    // so we pad it with a char that is not part of the B64 char set to prevent wrongful detection
    // and then remove it later.
    const bal = argv.indexOf("balance") + 1;
    if (bal != 0 && argv[bal] && /-{1}[a-z0-9_-]{42}/i.test(argv[bal])) {
        balpad = true;
        argv[bal] = "[" + argv[bal];
    }
    // padding hack to wallet addresses as well
    const wal = (!argv.includes("-w") ? argv.indexOf("--wallet") : argv.indexOf("-w")) + 1;
    if (wal != 0 && argv[wal] && /-{1}.*/i.test(argv[wal])) {
        walpad = true;
        argv[wal] = "[" + argv[wal];
    }
    // pass the CLI our argv
    exports.program.parse(argv);
}
exports.exportForTesting = {
    path: __filename,
};
