// @irys/sdk/build/cjs/web/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseWebIrys = void 0;
const tslib_1 = require("tslib");
require("../common/hack.js");
const api_1 = tslib_1.__importDefault(require("../common/api"));
const fund_1 = tslib_1.__importDefault(require("../common/fund"));
const irys_1 = tslib_1.__importDefault(require("../common/irys"));
const utils_1 = tslib_1.__importDefault(require("../common/utils"));
const provenance_1 = require("../common/provenance");
const transactions_1 = require("../common/transactions");
const arbundles = tslib_1.__importStar(require("./utils"));
const upload_1 = require("./upload");
const approval_1 = require("../common/approval");
class BaseWebIrys extends irys_1.default {
    constructor({ url, network, wallet, config, getTokenConfig, }) {
        var _a, _b;
        // @ts-expect-error types
        super({ url, network, arbundles });
        this.api = new api_1.default({
            url: this.url,
            timeout: (_a = config === null || config === void 0 ? void 0 : config.timeout) !== null && _a !== void 0 ? _a : 100000,
            headers: config === null || config === void 0 ? void 0 : config.headers,
        });
        this.tokenConfig = getTokenConfig(this);
        if (this.url.host.includes("devnet.irys.xyz") && !((config === null || config === void 0 ? void 0 : config.providerUrl) || (wallet === null || wallet === void 0 ? void 0 : wallet.rpcUrl) || ((_b = this === null || this === void 0 ? void 0 : this.tokenConfig) === null || _b === void 0 ? void 0 : _b.inheritsRPC)))
            throw new Error(`Using ${this.url.host} requires a dev/testnet RPC to be configured! see https://docs.irys.xyz/developer-docs/using-devnet`);
        this.token = this.tokenConfig.name;
        this.utils = new utils_1.default(this.api, this.token, this.tokenConfig);
        this.uploader = new upload_1.WebUploader(this);
        this.funder = new fund_1.default(this.utils);
        this.uploader = new upload_1.WebUploader(this);
        this.provenance = new provenance_1.Provenance(this);
        this.transactions = new transactions_1.Transaction(this);
        this.approval = new approval_1.Approval(this);
        this.address = "Please run `await Irys.ready()`";
        this.uploadFolder = this.uploader.uploadFolder.bind(this.uploader);
        this.uploadFile = this.uploader.uploadFile.bind(this.uploader);
    }
}
exports.BaseWebIrys = BaseWebIrys;
