// @irys/sdk/build/cjs/common/irys.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const transaction_1 = tslib_1.__importDefault(require("./transaction"));
const utils_1 = tslib_1.__importDefault(require("./utils"));
const withdrawal_1 = require("./withdrawal");
const query_1 = tslib_1.__importDefault(require("@irys/query"));
class Irys {
    constructor({ url, network, arbundles }) {
        switch (network) {
            // case undefined:
            case "mainnet":
                url = "https://arweave.mainnet.irys.xyz";
                break;
            case "devnet":
                url = "https://arweave.devnet.irys.xyz";
                break;
        }
        if (!url)
            throw new Error(`Missing required Irys constructor parameter: URL or valid Network`);
        const parsed = new URL(url);
        this.url = parsed;
        this.arbundles = arbundles;
        this.IrysTransaction = (0, transaction_1.default)(this);
    }
    get address() {
        if (!this._address)
            throw new Error("Address is undefined, please provide a wallet or run `await irys.ready()`");
        return this._address;
    }
    set address(address) {
        this._address = address;
    }
    get signer() {
        return this.tokenConfig.getSigner();
    }
    get search() {
        const q = new query_1.default({ url: new URL("/graphql", this.url) });
        return q.search.bind(q);
    }
    query(queryOpts) {
        return new query_1.default(queryOpts !== null && queryOpts !== void 0 ? queryOpts : { url: new URL("graphql", this.url) });
    }
    withdrawBalance(amount) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return (0, withdrawal_1.withdrawBalance)(this.utils, this.api, amount);
        });
    }
    withdrawAll() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return (0, withdrawal_1.withdrawBalance)(this.utils, this.api, "all");
        });
    }
    /**
     * Gets the balance for the loaded wallet
     * @returns balance (in winston)
     */
    getLoadedBalance() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!this.address)
                throw new Error("address is undefined");
            return this.utils.getBalance(this.address);
        });
    }
    /**
     * Gets the balance for the specified address
     * @param address address to query for
     * @returns the balance (in winston)
     */
    getBalance(address) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return this.utils.getBalance(address);
        });
    }
    /**
     * Sends amount atomic units to the specified bundler
     * @param amount amount to send in atomic units
     * @returns details about the fund transaction
     */
    fund(amount, multiplier) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return this.funder.fund(amount, multiplier);
        });
    }
    /**
     * Calculates the price for [bytes] bytes for the loaded token and Irys node.
     * @param bytes
     * @returns
     */
    getPrice(bytes) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return this.utils.getPrice(this.token, bytes);
        });
    }
    verifyReceipt(receipt) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return utils_1.default.verifyReceipt(this.arbundles, receipt);
        });
    }
    /**
     * Create a new IrysTransactions (flex token arbundles dataItem)
     * @param data
     * @param opts - dataItemCreateOptions
     * @returns - a new IrysTransaction instance
     */
    createTransaction(data, opts) {
        return new this.IrysTransaction(data, this, opts);
    }
    /**
     * Returns the signer for the loaded token
     */
    getSigner() {
        return this.tokenConfig.getSigner();
    }
    upload(data, opts) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return this.uploader.uploadData(data, opts);
        });
    }
    /**
     * @deprecated - use upload instead
     */
    uploadWithReceipt(data, opts) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return this.uploader.uploadData(data, Object.assign({}, opts));
        });
    }
    ready() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.tokenConfig.ready ? yield this.tokenConfig.ready() : true;
            this.address = this.tokenConfig.address;
            return this;
        });
    }
    // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
    get transaction() {
        // eslint-disable-next-line @typescript-eslint/no-this-alias
        const oThis = this;
        return {
            fromRaw(rawTransaction) {
                return new oThis.IrysTransaction(rawTransaction, oThis, { dataIsRawTransaction: true });
            },
        };
    }
}
Irys.VERSION = "0.2.5";
exports.default = Irys;
